src/build_helper/src/git.rs

```diff
@@ -176,44 +186,65 @@ fn get_latest_commit_that_modified_files(
     if output.is_empty() { Ok(None) } else { Ok(Some(output)) }
 }
 
-/// Returns the most recent commit found in the local history that should definitely
-/// exist upstream. We identify upstream commits by the e-mail of the commit author.
+/// Returns the most recent (ordered chronologically) commit found in the local history that
+/// should exist upstream. We identify upstream commits by the e-mail of the commit
+/// author.
 ///
-/// If `include_head` is false, the HEAD (current) commit will be ignored and only
-/// its parents will be searched. This is useful for try/auto CI, where HEAD is
-/// actually a commit made by bors, although it is not upstream yet.
+/// If we are in CI, we simply return out first parent.
 fn get_closest_upstream_commit(
     git_dir: Option<&Path>,
     config: &GitConfig<'_>,
     env: CiEnv,
 ) -> Result<Option<String>, String> {
+    let base = match env {
+        CiEnv::None => "HEAD",
+        CiEnv::GitHubActions => {
+            // On CI, we should always have a non-upstream merge commit at the tip,
+            // and our first parent should be the most recently merged upstream commit.
+            // We thus simply return our first parent.
+            return resolve_commit_sha(git_dir, "HEAD^1").map(Some);
+        }
+    };
+
     let mut git = Command::new("git");
 
     if let Some(git_dir) = git_dir {
         git.current_dir(git_dir);
     }
 
-    let base = match env {
-        CiEnv::None => "HEAD",
-        CiEnv::GitHubActions => {
-            // On CI, we always have a merge commit at the tip.
-            // We thus skip it, because although it can be created by
-            // `config.git_merge_commit_email`, it should not be upstream.
-            "HEAD^1"
-        }
-    };
+    // We do not use `--first-parent`, because we can be in a situation (outside CI) where we have
+    // a merge of a subtree that actually has the main rustc history as its second parent.
+    // Using `--first-parent` would recurse into the history of the subtree, which could have some
+    // old bors commits that are not relevant to us.
+    // Without `--first-parent`, git recurses into all parent subtrees. We thus need to decide which
+    // bors parent to choose (again, the subtree can have some old bors commits).
+    // Since bors is now used only for rust-lang/rust, we use a simple heuristic that selects the
+    // bors commit that was committed most recently (hence `--author-date-order`).
```

Comment:
FWIW at this point any topological sorting should do (which I assume git does by default), i.e., printing child commits before parent commits and using an arbitrary order for unrelated commits -- the bors commits in Miri are so old that they all have rustc bors commits as descendents.

Reply:
I'm not sure if it actually uses the author date or the committer date. In theory bors could recently merge a PR with a commit that was authored a long time ago. So I would rather explicitly ask for `--author-date-order`.

Reply:
bors is both the author and the committer, so having the two dates be different should be impossible?

Reply:
Ah, I guess you're right, I was simulating this situation in my tests and experiments, but it probably won't happen with actual bors. I would still prefer to use the explicit option than the default for which it is not clearly documented what it does (or at least I haven't found it).